### PR TITLE
Provenance

### LINKED ISSUE #81: Provenance for Whisper

### PR DESCRIPTION
**Pull request by greenw0lf**
Closes #81

### CODE REVIEW

File: asr.py [hunk 1]
```diff
@@ -11,50 +25,111 @@
...
     if not asr_already_done(output_path):
         logger.info("No Whisper transcript found")
-        run_asr(input_path, output_path, model)
+        whisper_prov = run_asr(input_path, output_path, model)
+        if whisper_prov:
+            prov_steps.append(whisper_prov)
     else:
         logger.info(f"Whisper transcript already present in {output_path}")
+        provenance = {
+            "activity_name": "Whisper transcript already exists",
```
Comment (mwigham) [inline asr.py:68]: interesting point for later discussion with the team as to whether we model this as a different activity, or as a different output of the Whisper speech processing activity. There's something to be said for both

File: transcode.py [hunk 2]
```diff
@@ -1,22 +1,47 @@
...
+def try_transcode(input_path, asset_id, extension) -> Optional[TranscodeOutput]:
     logger.info(
         f"Determining if transcode is required for input_path: {input_path} asset_id: ({asset_id}) extension: ({extension})"
     )
+    start_time = time.time()
+
+    provenance = {
+        "activity_name": "Transcoding",
+        "activity_description": "Checks if input needs transcoding, then transcodes if so",
+        "processing_time_ms": 0,
```
Comment (mwigham) [inline transcode.py:28]: I would personally prefer something impossible like -1 as a default, so it is clear what the difference is between an actual measurement and a default value.

File: transcode.py [hunk 3]
```diff
@@ -1,22 +1,47 @@
...
+    }
 
     # if it's alrady valid audio no transcode necessary
     if _is_audio_file(extension):
         logger.info("No transcode required, input is audio")
-        return input_path
+        end_time = (time.time() - start_time) * 1000
+        provenance["processing_time_ms"] = end_time
+        provenance["output_data"] = input_path
+        provenance["steps"].append("No transcode required, input is audio")
```
Comment (mwigham) [inline transcode.py:43]: Here again we need to think (in the future) a bit more clearly as a team about how we want to communicate about steps that were skipped (but not failed).

File: asr.py [hunk 4]
```diff
@@ -11,50 +25,111 @@
 
 logger = logging.getLogger(__name__)
 os.environ["HF_HOME"] = model_base_dir  # change dir where model is downloaded
+my_version = pkg_resources.get_distribution(
```
Comment (mwigham) [inline asr.py:28]: This is really interesting. How does this work to get the worker version? We have struggled previously with trying to put in a Github version but we had trouble getting that into the Dockerfile.

**Review by mwigham — APPROVED** [review]
Great!
I like it very much that each step returns the provenance in its output
I also really like the inclusion of config elements as parameters

Some of your choices in modelling the provenance are interesting (and I mean that in a positive way). Something for later discussion in the team.

As far as I'm concerned this is very good for the PoC. You might want to check with @jblom as to whether all the elements that he is interested in about the worker are included.

File: asr.py (hunk 4 above)
Comment (greenw0lf) [reply asr.py:28]: This approach doesn't put in the Github version. It simply obtains the version from the `pyproject.toml` file. And also, this approach wasn't actually working, so I'm replacing it with a different one that actually reads `pyproject.toml` and outputs what is written in the version field.

**Comment by jblom** [comment]
@greenw0lf @mwigham ok I've globally checked the code and comments here and it seems fine to merge it to the "dane as a service branch". From there I'll also test the code a bit more thoroughly and then merge with main, so we can try it out later this week.

**Pull request merged**